# Differential expression analysis

Once the RNAseq reads aligned on a reference and their per-gene abundance quantified, differential expression analysis can be conducted. Such analysis consists in identifying the genes that are over- and under-expressed in the test condition, compared to the control condition.
Here, we will try to identify the genes that are differentially expressed in the Flg22 mutants, compared to the Col0 plants. 

## 0. Prior to the differential expression analysis

Create a working directory and link the mapping files we will use in it:

```
mkdir ~/DGE_session
ln -s /netscratch/common/MPIPZ_SPP_workshop/RNA-Seq/RNA-Seq_align_data/ ~/DGE_session/

```

## 1. Importing mapping outputs in R

The R package `tximport` has been developed to easily import data generated by different mappers (including Kallisto), prior to a differential expression analysis. 

By having a look at the Kallisto output files (in folder ```~/DGE_session/RNA-Seq_align_data/Col0_s_rep1_ath_kallisto/``` for instance), you can observe that two types of files have been generated:
- .tsv files contain per-gene read counts and associated statistics such as gene length. Have a look at one of them using the following command: `head -n 10 abundance.tsv`
- .h5 files are binary files that contain the exact same information. They are not 'readable', but way lighter. Using these files rather than the .tsv ones will make your analysis more memory-efficient.

Now let's use R.
Let's open the `abundance.h5` files for the three replicates of our two experimental conditions, using tximport.

```r

# First, we import the package 'tximport'
library(tximport)

# Let's create a vector containing the names of the files we want to import
setwd('~/DGE_session')
files <- c(
     'RNA-Seq_align_data/Col0_s_rep1_ath_kallisto/abundance.h5',
     'RNA-Seq_align_data/Col0_s_rep2_ath_kallisto/abundance.h5',
     'RNA-Seq_align_data/Col0_s_rep3_ath_kallisto/abundance.h5',
     'RNA-Seq_align_data/flg22_s_rep1_ath_kallisto/abundance.h5',
     'RNA-Seq_align_data/flg22_s_rep2_ath_kallisto/abundance.h5',
     'RNA-Seq_align_data/flg22_s_rep3_ath_kallisto/abundance.h5' )

# We open a file that is needed by tximport to work (mapping between transcriptIDs and geneIDs)
tx2gene <- read.csv(file="tx2gene", header=TRUE, sep=",")

# Now, we can load the abundance.h5 files
txi.kallisto.h5 <- tximport(files, type = "kallisto", tx2gene = tx2gene, ignoreTxVersion = TRUE)

# Let's tell R which dataset correspond to which condition
sampleTable <- data.frame(condition = factor(c("Col0","Col0","Col0","Flg22","Flg22","Flg22")))
     
```

## 2. Performing the differential expression analysis with DESeq2

Multiple R packages dedicated to differential expression analyses exist. The most used are DESeq2, edgeR and Limma. Although they all have specificities, their main goal is to perform the following tasks:
- calculate log2FoldChanges values between the test and control conditions
- take into consideration the total amount of reads in each sample (sequencing depth)
- test statistically the differences in read counts between the two conditions, for each gene

```r

# Load DESeq2 and link it to the data imported with tximport
library(DESeq2)
data <- DESeqDataSetFromTximport(txi.kallisto.h5, sampleTable, ~condition)

# Tell DESeq which condition is the control one
data$condition <- relevel(data$condition, ref = "Col0")

# Run DESeq
analysis <- DESeq(data)

# Get results from the DESeq analysis, then sort them by p-value
res <- results(analysis)
res <- res[order(res$padj),]

# Have a look at what they look like
head(res)

# Now, retrieve the list of differentially-expressed genes
resSig <- subset(res, padj < 0.05)
resOverexp <- subset(resSig, log2FoldChange > 0)
resUnderexp <- subset(resSig, log2FoldChange < 0)

```

We could stop the analysis here. However, we might want to use the log2FoldChange values in a latter analysis.
You can already see log2FoldChange values in the 'res' dataframes. However, these are biased. Indeed, a gene that has a low expression (and therefore small read counts) is more likely to get a very high log2FoldChange. This would not necessarily mean that this gene is over-expressed in the test condition. 
Methods have been developed to shrink these log2FoldChange values. Here, we will use `apeglm` (Zhu et al., 2018).


```r

# Get results with shrinked Log2FoldChange values
resLFC <- lfcShrink(analysis, coef="condition_Flg22_vs_Col0", type="apeglm")

# Plot the log2FCs
plotMA(resLFC, ylim=c(-2,2))

```

From the statistical testing performed by DESeq2, we could analyse gene by gene the transcriptomic differences between our two conditions. However, we will try to get a global overview of these differences.

## 3. GO Enrichment Analysis

Gene Ontology is a hierarchical functional classification of genes. It allows to tell if a specific function is enriched in a gene-set of interest. 
A common application of this method consists in looking for a GO term enrichment in the list of over- or under-expressed genes retrieved by DESeq. 

As we are working on **Arabidopsis thaliana** - a model organism -, we can use dedicated online tools. For instance, we can retrieve the list of overexpressed genes from our previous analysis (```toString(overexpGenes)```), and then, copy-paste it in the text-box of the following website: https://www.arabidopsis.org/tools/go_term_enrichment.jsp

Another way - that can be used on any organism - consist in using a R package such as `topGO`. 

GO terms are divided in three categories: Biological Process, Molecular Function and Cellular Component. We will check for enrichment in GO terms of the two first categories using topGO:


```r

library(topGO)

# Let's load the association of each Arabidopsis gene to GO terms 
gene2go <- readMappings(file='gene2go', IDsep=',')

# Here, we get the gene names of all the genes from our data set, and in another vector, the ones that are overexpressed
allGenes <- rownames(res)
overexpGenes <- rownames(resOverexp)

# Let's format our geneset so it can be accepted by topGO
geneList <- factor(as.integer(allGenes %in% overexpGenes))
names(geneList) <- allGenes

# Now we can create a topGO object for Molecular Function GO category, then run a Classic Fisher Test 
mfGOdata <- new("topGOdata", ontology = "MF", allGenes = geneList, annot = annFUN.gene2GO, gene2GO = gene2go)
allGO <- usedGO(object = mfGOdata) 
mfFisher <- runTest(mfGOdata, algorithm = "classic", statistic = "fisher", scoreOrder = "increasing")

# Retrieve the results and correct the P-values with a BH method
mfRes <- GenTable(mfGOdata,classic = mfFisher, topNodes = length(allGO))
fdr<-p.adjust(mfRes$classic, "BH")
mfRes <-cbind(mfRes,fdr)

# See the results
subset(mfRes, fdr<0.05)
           
# Now let's do the same on the Biological Process GO category
bpGOdata <- new("topGOdata", ontology = "BP", allGenes = geneList, annot = annFUN.gene2GO, gene2GO = gene2go)
allGO <- usedGO(object = bpGOdata) 
bpFisher <- runTest(bpGOdata, algorithm = "classic", statistic = "fisher", scoreOrder = "increasing")
bpRes <- GenTable(bpGOdata,classic = bpFisher, topNodes = length(allGO))
fdr<-p.adjust(bpRes$classic, "BH")
bpRes <-cbind(bpRes,fdr)


```

